dxr/vcs.py

```diff
@@ -0,0 +1,338 @@
+        # doesn't seem to do that nicely
+        return self.upstream + "/commit/" + self.revision
+
+    def generate_blame(self, path):
+        return self.upstream + "/blame/" + self.revision + "/" + path
+
+    def generate_log(self, path):
+        return self.upstream + "/commits/" + self.revision + "/" + path
+
+    def get_contents(self, path, revision):
+        return self.invoke_vcs(['show', revision + ':' + path])
+
+class Perforce(Vcs):
+    def __init__(self, root, upstream):
+        super(Perforce, self).__init__(root, 'p4')
+        have = self._p4run(['have'])
+        self.have = dict((x['path'][len(root) + 1:], x) for x in have)
+        self.upstream = upstream
+
+    @classmethod
+    def claim_vcs_source(cls, path, dirs, tree):
+        if 'P4CONFIG' not in os.environ:
+            return None
+        if os.path.exists(os.path.join(path, os.environ['P4CONFIG'])):
+            return cls(path, tree.p4web_url)
+        return None
+
+    def _p4run(self, args):
+        ret = []
+        env = os.environ
+        env["PWD"] = self.root
+        proc = subprocess.Popen(['p4', '-G'] + args,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            cwd=self.root,
+            env=env)
+        while True:
+            try:
+                x = marshal.load(proc.stdout)
+            except EOFError:
+                break
+            ret.append(x)
+        return ret
+
+    def is_tracked(self, path):
+        return path in self.have
+
+    def generate_raw(self, path):
+        info = self.have[path]
+        return self.upstream + info['depotFile'] + '?ac=98&rev1=' + info['haveRev']
+
+    def generate_diff(self, path):
+        info = self.have[path]
+        haveRev = info['haveRev']
+        prevRev = str(int(haveRev) - 1)
+        return (self.upstream + info['depotFile'] + '?ac=19&rev1=' + prevRev +
+                '&rev2=' + haveRev)
+
+    def generate_blame(self, path):
+        info = self.have[path]
+        return self.upstream + info['depotFile'] + '?ac=193'
+
+    def generate_log(self, path):
+        info = self.have[path]
+        return self.upstream + info['depotFile'] + '?ac=22#' + info['haveRev']
+
+    def display_rev(self, path):
+        info = self.have[path]
+        return '#' + info['haveRev']
+
+every_vcs = [Mercurial, Git, Perforce]
+
+def tree_to_repos(tree):
+    """Given a TreeConfig, return a mapping {root: Vcs object} where root is a
+    directory under tree.source_folder where root is a directory under
+    tree.source_folder. Traversal of the returned mapping follows the order of
+    deepest directory first.
+
+    :arg tree: TreeConfig object representing a source code tree
+
+    """
+    sources = {}
+    # Find all of the VCSs in the source directory:
+    # We may see multiple VCS if we use git submodules, for example.
+    for cwd, dirs, files in os.walk(tree.source_folder):
```

Comment:
Also the permalink is created at index time; if we want to have it not appear in case the source is not available, we can move it to a skimmer object and do it at request time (by passing the VCS cache into the FileToSkim constructor).